src/System.Memory/src/System/Buffers/SequenceReader_search.cs

```diff
@@ -0,0 +1,798 @@
+        /// <param name="delimiter">The delimiter to look for.</param>
+        /// <returns>True if the <paramref name="delimiter"/> was found.</returns>
+        public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter)
+        {
+            ReadOnlySpan<T> remaining = UnreadSpan;
+            int index = remaining.IndexOf(delimiter);
+
+            if (index != -1)
+            {
+                span = remaining.Slice(0, index);
+                AdvanceCurrentSpan(index + 1);
+                return true;
+            }
+
+            return TryReadToMultisegment(out span, delimiter);
+        }
+
+        private bool TryReadToMultisegment(out ReadOnlySpan<T> span, T delimiter)
+        {
+            return TryReadToSlow(out span, delimiter, advancePastDelimiter: true);
+        }
+
+        public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter, bool advancePastDelimiter)
+        {
+            ReadOnlySpan<T> remaining = UnreadSpan;
+            int index = remaining.IndexOf(delimiter);
+
+            if (index != -1)
+            {
+                span = index == 0 ? default : remaining.Slice(0, index);
+                AdvanceCurrentSpan(index + (advancePastDelimiter ? 1 : 0));
+                return true;
+            }
+
+            return TryReadToSlow(out span, delimiter, advancePastDelimiter);
+        }
+
+        private bool TryReadToSlow(out ReadOnlySpan<T> span, T delimiter, bool advancePastDelimiter)
+        {
+            if (!TryReadToInternal(out ReadOnlySequence<T> sequence, delimiter, advancePastDelimiter, CurrentSpan.Length - CurrentSpanIndex))
+            {
+                span = default;
+                return false;
+            }
+
+            span = sequence.IsSingleSegment ? sequence.First.Span : sequence.ToArray();
+            return true;
+        }
+
+        /// <summary>
+        /// Try to read everything up to the given <paramref name="delimiter"/>, ignoring delimiters that are
+        /// preceeded by <paramref name="delimiterEscape"/>.
+        /// </summary>
+        /// <param name="span">The read data, if any.</param>
+        /// <param name="delimiter">The delimiter to look for.</param>
+        /// <param name="delimiterEscape">If found prior to <paramref name="delimiter"/> it will skip that occurence.</param>
+        /// <param name="advancePastDelimiter">True to move past the <paramref name="delimiter"/> if found.</param>
+        /// <returns>True if the <paramref name="delimiter"/> was found.</returns>
+        public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter, T delimiterEscape, bool advancePastDelimiter = true)
+        {
+            ReadOnlySpan<T> remaining = UnreadSpan;
+            int index = remaining.IndexOf(delimiter);
+
+            if ((index > 0 && !remaining[index - 1].Equals(delimiterEscape)) || index == 0)
+            {
+                span = remaining.Slice(0, index);
+                Advance(index + (advancePastDelimiter ? 1 : 0));
```

Comment:
@GrabYourPitchforks, you were looking at optimizations around making Convert.ToInt32(bool) branchless.  Did anything there materialize?